Reduce allocation in Process.GetProcessesByName

stephentoub opened the pull request:
GetProcessesByName is loading all of the information for every process found, even if it'll immediately be thrown away because the name doesn't match.

```C#
[Benchmark]
public void GetProcessesByName()
{
    foreach (Process p in Process.GetProcessesByName("dotnet.exe"))
        p.Dispose();
}
```

|             Method |         Toolchain |     Mean | Ratio | Allocated | Alloc Ratio |
|------------------- |------------------ |---------:|------:|----------:|------------:|
| GetProcessesByName | \main\corerun.exe | 4.415 ms |  1.00 | 622.43 KB |       1.000 |
| GetProcessesByName |   \pr\corerun.exe | 4.066 ms |  0.91 |   2.14 KB |       0.003 |

danmoseley commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.FreeBSD.cs:
```diff
@@ -20,28 +20,32 @@ internal static string GetProcPath(int processId)
-
             // Try to get the task info. This can fail if the user permissions don't permit
             // this user context to query the specified process
             ProcessInfo iinfo = Interop.Process.GetProcessInfoById(pid);
```
why 'iinfo'

stephentoub replied:
no idea. I didn't write or change it here.

danmoseley commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs:
```diff
@@ -330,71 +330,63 @@ private static unsafe ProcessInfo[] GetProcessInfos(ReadOnlySpan<byte> data, int
+                            processId == NtProcessManager.IdleProcessID ? "Idle" :
+                            processId.ToString(CultureInfo.InvariantCulture)); // use the process ID for a normal process without a name
+
+                    if (string.IsNullOrEmpty(processNameFilter) || processNameSpan.SequenceEqual(processNameFilter))
```
should this be using case insensitive comparison?

stephentoub replied:
Added

danmoseley commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Windows.cs:
```diff
@@ -121,7 +140,10 @@ public static ProcessInfo[] GetProcessInfos(string machineName)
                 {
-                    return NtProcessInfoHelper.GetProcessShortName(processName);
+                    ReadOnlySpan<char> newName = NtProcessInfoHelper.GetProcessShortName(processName);
+                    return newName.SequenceEqual(processName) ?
```
case insensitive?

stephentoub replied:
No, that would be pure overhead

danmoseley replied:
Oops, misread. Does this suggest an API on `ReadOnlySpan<char>` like `string ToString(string unifywith)` or some better name..

stephentoub commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.UnknownUnix.cs:
```diff
@@ -12,6 +12,7 @@ public static int[] GetProcessIds()
         }
 
         /// <summary>Gets process infos for each process on the specified machine.</summary>
+        /// <param name="processNameFilter">Optional process name to use as an inclusion filter.</param>
```
I think this whole file can be deleted

danmoseley commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs:
```diff
@@ -271,7 +271,7 @@ internal static class NtProcessInfoHelper
         private static uint MostRecentSize = DefaultCachedBufferSize;
 
-        internal static unsafe ProcessInfo[] GetProcessInfos(int? processIdFilter = null)
+        internal static unsafe ProcessInfo[] GetProcessInfos(int? processIdFilter = null, string? processNameFilter = null)
```
should these document that the filters must both be satisfied?

stephentoub committed:
Reduce allocation in Process.GetProcessesByName

GetProcessesByName is loading all of the information for every process found, even if it'll immediately be thrown away because the name doesn't match.

danmoseley commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Linux.cs:
```diff
@@ -13,15 +13,18 @@ internal static partial class ProcessManager
             ThrowIfRemoteMachine(machineName);
 
             // Iterate through all process IDs to load information about each process
+            IEnumerable<int> pids = EnumerateProcessIds();
```
any particular reason for pulling this out into a variable?

stephentoub replied:
For symmetry with the corresponding bsd method.

danmoseley commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.OSX.cs:
```diff
@@ -40,10 +37,28 @@ private static ProcessInfo CreateProcessInfo(int pid)
+            }
+            else
+            {
+                procInfo = new ProcessInfo()
```
should these dummies be suppressed if there's a filter passed? that may be the cause of the test failure.

marek-safar commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.BSD.cs:
```diff
@@ -18,27 +18,24 @@ public partial class Process
-            Process[] procs = GetProcesses(machineName);
-            var list = new List<Process>();
+            int[] procIds = ProcessManager.GetProcessIds();
+            var processes = string.IsNullOrEmpty(processName) ?
+                new List<Process>(procIds.Length) :
+                new List<Process>();
```
What about always allocating Process[procIds.Length] and resize when there is null entry?

jkotas replied:
`ArrayBuilder` under Common does exactly that.

stephentoub committed:
Fix a few issues and address PR feedback

stephentoub commented:
Note this also fixed the unreported issue on Linux/macOS that passing null or empty string to GetProcessesByName shouldn't filter anything out but was filtering everything out.

Nuklon commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Linux.cs:
```diff
@@ -24,20 +24,21 @@ public partial class Process : IDisposable
-            }
 
-            var processes = new List<Process>();
+            processName ??= "";
```
Just curious: is there any reason for now using "" instead of string.Empty?

stephentoub replied:
Just consistency, e.g. a test using InlineData to feed inputs can't use string.Empty, and since I was adding additional uses, I made them all "".

danmoseley replied:
@Nuklon i don't believe there is ever a reason to use String.Empty. It is an anachronism.

Nuklon replied:
Thanks!
